### PR TITLE
Add ProductionCost/TimeMultiplier

### PR DESCRIPTION
**Pull request by MustaphaTR**
Applies https://github.com/OpenRA/OpenRA/pull/15305 to here.

I have edited this to be a multiplier to be less hacky tho haven't tested much yet, so it better be tested first. I don't have much time left for today.

### CODE REVIEW

**Comment by GraionDilach** [comment]
I haven't noticed this before I go to sleep but I'll try to check this out ASAP.

**Comment by MustaphaTR** [comment]
Don't merge this yet. As i said on the PR on OpenRA repo, i messed something up while trying to do this with interfaces, i'll change this so it'll use interfaces.

**Comment by GraionDilach** [comment]
You should use conditions instead of prerequsites IMO.

**Comment by MustaphaTR** [comment]
Production is done thru actor infos. Conditions won't work there.

**Comment by GraionDilach** [comment]
Sucks. A lot, actually.  This doesn't support stacking the same building for additional bonus without a bunch of complex yaml workarounds.

Then again, at this stage, the best solution I see would be to convert prereqs to booleanexpressions. but ofcourse that would be a horrible and nonsensical method of thinking... here.

Alright. I'll test this tonight. If it works as it should, I'll take it.

File: OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Modifies the production cost of this actor for a specific queue or when a prerequisite is granted.")]
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs]: -or

File: OpenRA.Mods.Common/Traits/Multipliers/ProductionTimeMultiplier.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Modifies the production time of this actor for a specific queue or when a prerequisite is granted.")]
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Multipliers/ProductionTimeMultiplier.cs]: Same here.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 3]
```diff
@@ -329,6 +329,8 @@
     <Compile Include="Traits\Conditions\GrantExternalConditionToCrusher.cs" />
     <Compile Include="Traits\Conditions\GrantRandomCondition.cs" />
     <Compile Include="Traits\Contrail.cs" />
+    <Compile Include="Traits\Multipliers\ProductionCostMultiplier.cs" />
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Only the interface should stay in Common namespace, the convention deployed here still prefers that traits be shifted to AS namespace. See ResourcePurifier for an example.

**Comment by GraionDilach** [comment]
The code itself looks okay, so it'll be merged after the fixup.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 3 above)
Comment (GraionDilach) [reply OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: This should also involve an Authors.AS edit thenafter.

File: OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs (hunk 1 above)
Comment (MustaphaTR) [reply OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs]: Queue is optional too, i feel like your sentence says it isn't.

Comment (GraionDilach) [reply OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs]: Point taken.

File: OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs [hunk 4]
```diff
@@ -138,8 +139,10 @@ void INotifyCreated.Created(Actor self)
 			// Created is called before Player.PlayerActor is assigned,
 			// so we must query other player traits from self, knowing that
 			// it refers to the same actor as self.Owner.PlayerActor
-			playerPower = (self.Info.Name == "player" ? self : self.Owner.PlayerActor).TraitOrDefault<PowerManager>();
-			productionTraits = self.TraitsImplementing<Production>().ToArray();
+			var playerActor = self.Info.Name == "player" ? self : self.Owner.PlayerActor;
+			playerPower = playerActor.TraitOrDefault<PowerManager>();
+			productionTraits = self.TraitsImplementing<Production>().Where(p => p.Info.Produces.Contains(Info.Type)).ToArray();
```
Comment (MustaphaTR) [inline OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs]: It seems like i'm applying a change from upstream that AS doesn't have yet. But it should break anything, should i revert that?

Comment (GraionDilach) [reply OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs]: No need, I'll do an engine update immediately after I merge this then.

**Comment by GraionDilach** [comment]
GJ. :+1:

**Pull request merged**